jinahub/indexers/dbms/PostgreSQLIndexer/postgreshandler.py

```diff
@@ -82,60 +98,64 @@ def use_table(self):
             except (Exception, Error) as error:
                 self.logger.error('Error while creating table!')
 
-    def add(self, ids, vecs, metas, *args, **kwargs):
-        """ Insert the documents into the database.
+    def add(self, docs: DocumentArray, *args, **kwargs):
+        """Insert the documents into the database.
 
-        :param ids: List of doc ids to be added
-        :param vecs: List of vecs to be added
-        :param metas: List of metas of docs to be added
+        :param docs: list of Documents
         :param args: other arguments
         :param kwargs: other keyword arguments
         :param args: other arguments
         :param kwargs: other keyword arguments
         :return record: List of Document's id added
         """
         row_count = 0
-        for i in range(len(ids)):
-            self.cursor.execute(
-                f'INSERT INTO {self.table} (ID, VECS, METAS) VALUES (%s, %s, %s)',
-                (ids[i], vecs[i].tobytes(), metas[i]),
-            )
-            row_count += self.cursor.rowcount
+        for doc in docs:
+            try:
+                self.cursor.execute(
+                    f'INSERT INTO {self.table} (ID, VECS, METAS) VALUES (%s, %s, %s)',
+                    (doc.id, doc.embedding.tobytes(), doc_without_embedding(doc)),
+                )
+                row_count += self.cursor.rowcount
+            except psycopg2.errors.UniqueViolation:
+                self.logger.warning(
+                    f'Document with id {doc.id} already exists in PSQL database. Skipping...'
+                )
+                self.connection.rollback()
         self.connection.commit()
```

Comment:
Just out of curiosity: Is anything send to the PSQL database before this `commit()` is called? If not, I am happy :)

Reply:
I will delve deeper into connection optimization in another PR. This one is growing quite large, as the integration tests and CI need to be running